File: 05/Java/TripleExpression/BinaryOperator.java
package TripleExpression;

abstract public class BinaryOperator implements TripleExpression {
    protected TripleExpression fOp;
    protected TripleExpression sOp;
    
    BinaryOperator(TripleExpression first, TripleExpression second) {
        fOp = first;
        sOp = second;
    }

    abstract protected double apply(double a, double b);
    
    public double evaluate(double x, double y, double z) {
        return apply(fOp.evaluate(x, y, z), sOp.evaluate(x, y, z));
    }
}

File: 08/CSharp/TripleExpressionApp/TripleExpressionApp/BinaryOperator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TripleExpressionApp
{
    abstract public class BinaryOperator<T> : ITripleExpression <T>
    {
        protected ITripleExpression<T> fOp;
        protected ITripleExpression<T> sOp;
        protected IOperator<T> op;

        public BinaryOperator(ITripleExpression<T> first, ITripleExpression<T> second, IOperator<T> op)
        {
            fOp = first;
            sOp = second;
            this.op = op;
        }
        public BinaryOperator() { }

        abstract protected T Apply(T a, T b);
        public T Evaluate(T x, T y, T z)
        {
            return Apply(fOp.Evaluate(x, y, z), sOp.Evaluate(x, y, z));
        }
    }
}
